skunkworks_crow/src/main/java/org/odk/share/views/ui/bluetooth/BtSenderActivity.java

```diff
@@ -148,18 +159,68 @@ private Disposable addUploadEventSubscription() {
     /**
      * Enable the bluetooth discovery for other devices. The timeout is specific seconds.
      */
-    private void enableDiscovery() {
+    @NeedsPermission({Manifest.permission.ACCESS_FINE_LOCATION,
+            Manifest.permission.ACCESS_COARSE_LOCATION})
+    void enableDiscovery() {
         Intent discoverableIntent = new Intent(BluetoothAdapter.ACTION_REQUEST_DISCOVERABLE);
         // set the discovery timeout for 120s.
         discoverableIntent.putExtra(BluetoothAdapter.EXTRA_DISCOVERABLE_DURATION, CONNECT_TIMEOUT);
-        startActivity(discoverableIntent);
+        startActivityForResult(discoverableIntent, DISCOVERABLE_CODE);
+    }
+
+    /**
+     * If the permission was denied, finishing this activity.
+     */
+    @OnPermissionDenied({Manifest.permission.ACCESS_FINE_LOCATION,
+            Manifest.permission.ACCESS_COARSE_LOCATION})
+    void showDeniedForLocation() {
+        Toast.makeText(this, R.string.permission_location_denied, Toast.LENGTH_LONG).show();
+        finish();
+    }
+
+    /**
+     * If clicked the "never ask", we should show a toast to guide user.
+     */
+    @OnNeverAskAgain({Manifest.permission.ACCESS_FINE_LOCATION,
+            Manifest.permission.ACCESS_COARSE_LOCATION})
+    void showNeverAskForLocation() {
+        PermissionUtils.showAppInfo(this, getPackageName());
+    }
+
+    @Override
+    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
+        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
+        BtSenderActivityPermissionsDispatcher.onRequestPermissionsResult(this, requestCode, grantResults);
+    }
+
+    @Override
+    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
+        super.onActivityResult(requestCode, resultCode, data);
+        switch (requestCode) {
+            case DISCOVERABLE_CODE:
+                if (resultCode == SUCCESS_CODE) {
+                    startCheckingDiscoverableDuration();
+                } else {
+                    finish();
+                }
+                break;
+            case APP_SETTING_REQUEST_CODE:
+                //only android 9 would enter this case.
```

Comment:
@huangyz0918 Please remove this comment, now every device with API >= 21 will execute this block.

Reply:
Sure! do this right now.